cpacker/MemGPT
# feat: CLI command to reload functions and system

*Please describe the purpose of this pull request.*
This is a new feature to support /reload functions and /reload system_prompt. It enables users to add or remove functions from an existing agent.

To add a function to an agent, the user must author a new function AND edit the preset for the agent

*How to test*
How can we test your PR during review? What commands should we run? What outcomes should we expect?

I didn't see a great way to add an actual test for this. But I've tested the following cases:

- new function added: agent able to call the function
- function removed: agent unable to call the function (no exception)
- function schema changed: at first I implemented this, but found that memgpt refuses to start if the schema is changed anyway, so I do not handle it.
- new function added that does not have an implementation: debug message but do not throw error.

Have you tested this PR?
I have tested the above use cases.

*Related issues or PRs*
https://github.com/cpacker/MemGPT/issues/577

*Is your PR over 500 lines of code?*
No

*Additional context*

Some open questions:

- should we add a system message to the agent to notify them that their function list has changed? I have left this off for now, but the feature is not all that usable, in my usage the agent is somewhat reluctant to try a new function
- should we document this feature in the README? Or leave for power users to discover?
- I'm wondering if there'd be a good way to test out alterations of the configs, etc. I think I saw somewhere that there's a major refactor coming that puts everything in the database, so not sure it'd be worth it
